### PR TITLE
fix-dynamo-subclass

### LINKED ISSUE #167635: TorchDynamo Compilation Error: Invalid Class Type in Custom Tensor Subclass During Gradient Computation
### 🐛 Describe the bug

Description：
    When compiling a model containing custom Tensor subclasses using torch.compile, a type conversion error occurred.
    A MySensor subclass was defined in the code, and methods such as sum, mean, and prod were rewritten. In the forward propagation of the model, transform a regular Tensor into this subclass and perform the corresponding operations. Although the model runs normally in normal mode, when using torch.compile, the internal type system cannot correctly handle custom Tensor subclasses, resulting in type conversion errors. Specifically, during backpropagation, the compiled model fails to correctly convert MySensor subclasses back to standard Tensor types, resulting in errors.
code:
```
import torch
import torch.nn as nn
import torch.nn.functional as F

class MyTensor(torch.Tensor):
    def sum(self, dim=None, keepdim=False):
        return super().sum(dim=dim, keepdim=keepdim).as_subclass(MyTensor)
    
    def mean(self, dim=None, keepdim=False):
        return super().mean(dim=dim, keepdim=keepdim).as_subclass(MyTensor)
    
    def prod(self, dim=None, keepdim=False):
        return super().prod(dim=dim, keepdim=keepdim).as_subclass(MyTensor)

class TestModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(10, 5)
    
    def forward(self, x):
        out = self.fc(x)
        out_subclass = out.as_subclass(MyTensor)
        reduced = out_subclass.sum(dim=1)
        return reduced

def get_default_model():
    return TestModel()

def get_sample_inputs():
    x = torch.randn(4, 10, requires_grad=True)
    return (x,)

def main():
    model = get_default_model()
    inputs = get_sample_inputs()
    x = inputs[0]
    output = model(x)
    target = torch.zeros_like(output)
    loss = F.mse_loss(output, target)
    loss.backward()
    print('TestModel executed successfully!')
    print('Input shape:', x.shape)
    print('Output shape:', output.shape)
    print('Model parameters count:', sum(p.numel() for p in model.parameters()))
    print('Input gradients shape:', x.grad.shape)
    try:
        compiled_model = torch.compile(model)
        with torch.no_grad():
            output_compile = compiled_model(*inputs)

        print(f'Compile  shape: {output_compile.shape}')
    except Exception as e:
        print(f'\nOriginal exception:\n {e}')

if __name__ == "__main__":
    main()
```
output:
```
TestModel executed successfully!
Input shape: torch.Size([4, 10])
Output shape: torch.Size([4])
Model parameters count: 55
Input gradients shape: torch.Size([4, 10])

Original exception:
 invalid class type <class '__main__.MyTensor'> in TensorWithTFOverrideVariable.from_tensor_var

from user code:
   File line 23, in forward
    reduced = out_subclass.sum(dim=1)
  File line 7, in sum
    return super().sum(dim=dim, keepdim=keepdim).as_subclass(MyTensor)
 line 1669, in __torch_function__
    ret = func(*args, **kwargs)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"

```

### Versions

PyTorch version: 2.8.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-6ubuntu2) 9.5.0
Clang version: Could not collect
CMake version: version 4.0.3
Libc version: glibc-2.39

Python version: 3.9.7 (default, Jul 16 2025, 16:34:47) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-29-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 580.65.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 39 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Core(TM) i9-14900HX
CPU family: 6
Model: 183
Thread(s) per core: 2
Core(s) per socket: 24
Socket(s): 1
Stepping: 1
CPU(s) scaling MHz: 31%
CPU max MHz: 5800.0000
CPU min MHz: 800.0000
BogoMIPS: 4838.40
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 896 KiB (24 instances)
L1i cache: 1.3 MiB (24 instances)
L2 cache: 32 MiB (12 instances)
L3 cache: 36 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Ghostwrite: Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by parsshar-RH**
Fixes #167635

### Summary:

- Fixes compiled code produces correct output and maintains the subclass type.
- Matches eager mode behavior

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

### CODE REVIEW

**Review by mlazos — APPROVED** [review]
This is great! thanks

File: torch/_dynamo/variables/torch_function.py [hunk 1]
```diff
@@ -588,7 +588,9 @@ def from_tensor_var(
         # This simulates shallow-copying the tensor object.
         kwargs = dict(tensor_var.__dict__)
         input_tensor_type = kwargs.pop("class_type")
-        assert input_tensor_type in (torch.Tensor, torch.nn.Parameter), (
+        assert input_tensor_type in (torch.Tensor, torch.nn.Parameter) or issubclass(
+            input_tensor_type, torch.Tensor
```
Comment (Skylion007) [inline torch/_dynamo/variables/torch_function.py:592]: What about subclass of nn.Parameter?

Comment (parsshar-RH) [reply torch/_dynamo/variables/torch_function.py:592]: I think subclasses of `nn.Paramter` is covered by `issubclass(input_tensor_type, torch.Tensor)` since `nn.Parameter `inherits from `torch.Tensor`

Please feel free to suggest if i am missing anything!

Thanks for the review!

**Comment by parsshar-RH** [comment]
Hi @Skylion007 

Please review the PR
Thanks!

**Review by cyyever — APPROVED** [review]
